### PR TITLE
fix(windows): save actor name to disk and reload it on startup

### LINKED ISSUE #3519: bug(windows): Still shows "TODO" for actor name if the app signs in automatically
**Describe the bug**
[A clear and concise description of what the bug is.](https://github.com/firezone/firezone/blob/d7ee1ebe88331650ef23fa0cc1d740526a9dac0f/rust/windows-client/src-tauri/src/client/auth.rs#L78-L84)

**To Reproduce**
1. Sign in to the Firezone app
2. Quit the app
3. Run the app again
4. It will show your name as "TODO" in the tray menu

**Expected behavior**
Expected actor name from SSO to be saved to disk and reloaded automatically

image

**Screenshots / Logs**
image

**Platform (please complete the following information):**
 - Firezone Version d7ee1ebe883
 - OS: Windows 11

**Additional context**
Just need to save it in %LOCALAPPDATA%. Probably in `data` somewhere.

### PR DESCRIPTION
**Pull request by ReactorScram**
Closes #3519 

The actor name is saved in plain text in our AppData folder.

I did simple manual tests on my dev laptop:
- Restart the app, it reloads the actor name from disk
- Sign out of the app, the file is deleted from the disk

Caveats:
- If you un-install the app while signed in, the actor name will still be there until something cleans up AppData. I don't think the MSI cleans it up.
- If the token is present but the actor name is missing, I just treat it as if the token was missing and make the app be signed-out, since this will happen on dev systems until this PR has been merged for some time.
- If we try to delete the file and the file is NotFound, I just ignore it.

### CODE REVIEW

**Review by jamilbk — COMMENTED** [review]
It's possible the actor name could be blank, e.g. a service account or an IdP that syncs a blank user name. Would this fail gracefully in that case?

**Comment by ReactorScram** [comment]
@jamilbk When I'm parsing the callback, the `actor_name` is required to be there, but it should work with an empty string since it's just an opaque `String` after that, any valid UTF-8 should work.

https://github.com/firezone/firezone/blob/079b4e7f9a6b991cac5ae270964c8ddebd3cc1e6/rust/windows-client/src-tauri/src/client/deep_link.rs#L54-L85

If that's okay I'll just add test cases for it and it'll end up saving `""` to a file?

**Comment by jamilbk** [comment]
So I think that might work for user accounts, but let's check with @AndrewDryga to see if that's always the case.

**Comment by AndrewDryga** [comment]
As I remember we don't allow empty names but from a client perspective Id ignore it and make sure empty one doesn't crash anything

**Comment by ReactorScram** [comment]
@AndrewDryga I added some unit tests to show that it shouldn't crash when parsing an empty name or saving / loading from disk:

- URL parsing https://github.com/firezone/firezone/blob/427dabc866d0de947bd210277c165fef58672f50/rust/windows-client/src-tauri/src/client/deep_link.rs#L258
- Save and reload, this gets run with `actor_name` set to empty and then set to "Jane Doe" https://github.com/firezone/firezone/blob/427dabc866d0de947bd210277c165fef58672f50/rust/windows-client/src-tauri/src/client/auth.rs#L328-L351

**Review by jamilbk — COMMENTED** [review]
LGTM, will defer to @conectado's approval on the Rust bits.

I think it's safe to assume `actor_name` will at least be there for the sign in flow, at least for now.

File: rust/windows-client/src-tauri/src/client/auth.rs [hunk 1]
```diff
@@ -1,21 +1,34 @@
...
 #[derive(thiserror::Error, Debug)]
 pub(crate) enum Error {
+    #[error("`actor_name_path` has no parent, this should be impossible")]
+    ActorNamePathWrong,
+    #[error("`app_local_data_dir` failed")]
+    CantFindLocalAppDataFolder,
+    #[error("`create_dir_all` failed while writing `actor_name_path`")]
+    CreateDirAll(std::io::Error),
+    #[error("Couldn't delete actor_name from disk: {0}")]
+    DeleteActorName(std::io::Error),
```
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/auth.rs:21]: Now that this is growing, I'm considering, is there any reason we aren't using `anyhow` considering this is a binary crate?

File: rust/windows-client/src-tauri/src/client/auth.rs [hunk 2]
```diff
@@ -149,25 +174,48 @@ impl Auth {
...
+                if error.kind() == std::io::ErrorKind::NotFound {
+                    // It can happen with dev systems that actor_name.txt doesn't exist
+                    // even though the token is in the cred manager.
+                    // In that case just sign the app out, don't crash.
+                    return Ok(None);
+                } else {
+                    return Err(Error::ReadActorName(error));
+                }
+            }
+        };
```
Comment (conectado) [inline rust/windows-client/src-tauri/src/client/auth.rs]:
```suggestion
        let actor_name = match std::fs::read_to_string(actor_name_path()?) {
              Ok(x) => x,
              Err(error) if error.kind() == std::io::ErrorKind::NotFound => { return Ok(None); }
              Err(error) => { return Err(Error::ReadActorName(error)); }
        };
```

File: rust/windows-client/src-tauri/src/client/auth.rs (hunk 1 above)
Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/auth.rs:21]: I like that with `thiserror` none of the `From`s are implemented by default, so I can't just put `?` and forget to add a `.context`

Comment (ReactorScram) [reply rust/windows-client/src-tauri/src/client/auth.rs:21]: Also, although I've never seen `anyhow` crash, it does use unsafe internally and `thiserror` does not, last time I checked

**Review by conectado — APPROVED** [review]

File: rust/windows-client/src-tauri/src/client/auth.rs (hunk 1 above)
Comment (conectado) [reply rust/windows-client/src-tauri/src/client/auth.rs:21]: > Also, although I've never seen anyhow crash, it does use unsafe internally and thiserror does not, last time I checked

`anyhow` is one of the most widely used crates, we can assume it's safe

> I like that with thiserror none of the Froms are implemented by default, so I can't just put ? and forget to add a .context

With anyhow you can have precise control over the context given the code-path and I'm not sure if the overhead of maintaining this enum is worth having this automatically added.

**Pull request merged**